### PR TITLE
SONAR-10288 ignore empty fields in multivalue properties

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/config/MultivaluePropertyCleaner.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+   *    <li>{@code "a,\"\",b" => "a,b"}</li>
+   *    <li>{@code "\"a\",\"b\"" => "\"a\",\"b\""}</li>
+   *    <li>{@code "\"  a  \",\"b \"" => "\"  a  \",\"b \""}</li>
+   *    <li>{@code "\"a\",\"\",\"b\"" => "\"a\",\"\",\"b\""}</li>
+   *    <li>{@code "\"a\",\"  \",\"b\"" => "\"a\",\"  \",\"b\""}</li>
+   *    <li>{@code "\"  a,,b,c  \",\"d \"" => "\"  a,,b,c  \",\"d \""}</li>
+   *    <li>{@code "a,\"  \",b" => "ab"]}</li>
+   * </ul>
+   */
+  public static String trimFieldsAndRemoveEmptyFields(String str) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/config/MultivaluePropertyCleaner.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWD-IMYHEFfjxkQcPtnB&open=AWD-IMYHEFfjxkQcPtnB)

[ISSUE_KEY]: AWD-IMYHEFfjxkQcPtnB

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    return configuration;
+  }
+
+  private static class ServerConfigurationAdapter implements Configuration {
+    private static final Function<String, String> REPLACE_ENCODED_COMMAS = value -> value.replace("%2C", ",");
+
+    private final Settings settings;
+
+    private ServerConfigurationAdapter(Settings settings) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: Code Smell Code Smell: Remove this use of "Settings"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWD_mmyFEFfjxkQcP2WX&open=AWD_mmyFEFfjxkQcP2WX)

[ISSUE_KEY]: AWD_mmyFEFfjxkQcP2WX

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    if (configuration == null) {
+      configuration = new ServerConfigurationAdapter(settings);
+    }
+    return configuration;
+  }
+
+  private static class ServerConfigurationAdapter implements Configuration {
+    private static final Function<String, String> REPLACE_ENCODED_COMMAS = value -> value.replace("%2C", ",");
+
+    private final Settings settings;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: Code Smell Code Smell: Remove this use of "Settings"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWD_mmyFEFfjxkQcP2WW&open=AWD_mmyFEFfjxkQcP2WW)

[ISSUE_KEY]: AWD_mmyFEFfjxkQcP2WW

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import org.sonar.api.config.Settings;
+
+import static java.util.function.Function.identity;
+import static org.sonar.core.config.MultivalueProperty.parseAsCsv;
+
+public class ConfigurationProvider extends ProviderAdapter {
+
+  private Configuration configuration;
+
+  public Configuration provide(Settings settings) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: Code Smell Code Smell: Remove this use of "Settings"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWD_mmyFEFfjxkQcP2WV&open=AWD_mmyFEFfjxkQcP2WV)

[ISSUE_KEY]: AWD_mmyFEFfjxkQcP2WV

File: sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+   *    <li>{@code "\"a\",\"b\"" => "\"a\",\"b\""}</li>
+   *    <li>{@code "\"  a  \",\"b \"" => "\"  a  \",\"b \""}</li>
+   *    <li>{@code "\"a\",\"\",\"b\"" => "\"a\",\"\",\"b\""}</li>
+   *    <li>{@code "\"a\",\"  \",\"b\"" => "\"a\",\"  \",\"b\""}</li>
+   *    <li>{@code "\"  a,,b,c  \",\"d \"" => "\"  a,,b,c  \",\"d \""}</li>
+   *    <li>{@code "a,\"  \",b" => "ab"]}</li>
+   * </ul>
+   */
+  @VisibleForTesting
+  static String trimFieldsAndRemoveEmptyFields(String str) {
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWD_mm0-EFfjxkQcP2WY&open=AWD_mm0-EFfjxkQcP2WY)

[ISSUE_KEY]: AWD_mm0-EFfjxkQcP2WY

Comment (henryju) [inline sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java]: I wonder how far you are from totally replacing the CSVParser...

Comment (SonarTech) [reply sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java (hunk 4 above)
Comment (SonarTech) [reply server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java (hunk 3 above)
Comment (SonarTech) [reply server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java (hunk 5 above)
Comment (sns-seb) [reply sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java]: I know I would have to support escaping inside fields, may not be that complex, but that's overkill now.

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java (hunk 3 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: Code Smell Code Smell: Remove this use of "Settings"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWEuMi9xr8h_fVHZ4jCi&open=AWEuMi9xr8h_fVHZ4jCi)

[ISSUE_KEY]: AWEuMi9xr8h_fVHZ4jCi

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java (hunk 4 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: Code Smell Code Smell: Remove this use of "Settings"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWEuMi9xr8h_fVHZ4jCh&open=AWEuMi9xr8h_fVHZ4jCh)

[ISSUE_KEY]: AWEuMi9xr8h_fVHZ4jCh

File: sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java (hunk 5 above)
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. ([squid:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWEuMjBBr8h_fVHZ4jCk&open=AWEuMjBBr8h_fVHZ4jCk)

[ISSUE_KEY]: AWEuMjBBr8h_fVHZ4jCk

File: server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/config/ConfigurationProvider.java]: Code Smell Code Smell: Remove this use of "Settings"; it is deprecated. ([squid:CallToDeprecatedMethod](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10288%2Fignore_empty_fields_in_multivalue_properties&issues=AWEuMi9xr8h_fVHZ4jCj&open=AWEuMi9xr8h_fVHZ4jCj)

[ISSUE_KEY]: AWEuMi9xr8h_fVHZ4jCj

File: sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java (hunk 5 above)
Comment (SonarTech) [reply sonar-core/src/main/java/org/sonar/core/config/MultivalueProperty.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

**Pull request merged**